canhttp/src/observability/mod.rs

```diff
@@ -0,0 +1,241 @@
+    type Service = Observability<S, OnRequest, OnResponse, OnError>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Self::Service {
+            inner,
+            on_request: self.on_request.clone(),
+            on_response: self.on_response.clone(),
+            on_error: self.on_error.clone(),
+        }
+    }
+}
+
+/// Middleware that adds high level observability to a [`Service`].
+///
+/// See the [module docs](crate::observability) for an example.
+///
+/// [`Service`]: tower::Service
+pub struct Observability<S, OnRequest, OnResponse, OnError> {
+    inner: S,
+    on_request: OnRequest,
+    on_response: OnResponse,
+    on_error: OnError,
+}
+
+impl<S, Request, Response, OnRequest, RequestData, OnResponse, OnError> Service<Request>
+    for Observability<S, OnRequest, OnResponse, OnError>
+where
+    S: Service<Request, Response = Response>,
+    OnRequest: RequestObserver<Request, ObservableRequestData = RequestData>,
+    OnResponse: ResponseObserver<RequestData, S::Response> + Clone,
+    OnError: ResponseObserver<RequestData, S::Error> + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, RequestData, OnResponse, OnError>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        let req_data = self.on_request.observe_request(&req);
+        ResponseFuture {
+            response_future: self.inner.call(req),
+            request_data: Some(req_data),
+            on_response: self.on_response.clone(),
+            on_error: self.on_error.clone(),
+        }
+    }
+}
+
+/// Trait used to tell [`Observability`] what to do when a request is received.
+pub trait RequestObserver<Request> {
+    /// Type of data that can be observed from the request (e.g., URL, host, etc.)
+    /// when the response will be processed.
+    type ObservableRequestData;
+
+    /// Observe the given request and produce observable data based on the request.
+    /// This observable data will be passed on to the response observer.
+    fn observe_request(&self, request: &Request) -> Self::ObservableRequestData;
+}
+
+impl<Request> RequestObserver<Request> for () {
+    type ObservableRequestData = ();
+
+    fn observe_request(&self, _request: &Request) -> Self::ObservableRequestData {
+        //NOP
+    }
+}
+
+impl<F, Request, RequestData> RequestObserver<Request> for F
+where
+    F: Fn(&Request) -> RequestData,
+{
+    type ObservableRequestData = RequestData;
+
+    fn observe_request(&self, request: &Request) -> Self::ObservableRequestData {
+        self(request)
+    }
+}
+
+/// Trait used to tell [`Observability`] what to do when a response is received.
+pub trait ResponseObserver<RequestData, Result> {
+    /// Observe the response (typically an instance of [`std::Result`] and the request data produced by a [`RequestObserver`].
```

Comment:
nit: unmatched bracket
```suggestion
    /// Observe the response (typically an instance of [`std::Result`]) and the request data produced by a [`RequestObserver`].
```

Reply:
good catch, added.